server/schedule/region_splitter.go

```diff
@@ -165,28 +197,91 @@ func (h *splitRegionsHandler) SplitRegionByKeys(region *core.RegionInfo, splitKe
 	return nil
 }
 
-func (h *splitRegionsHandler) WatchRegionsByKeyRange(startKey, endKey []byte, splitKeys [][]byte, timeout, watchInterval time.Duration) map[uint64]struct{} {
+func (h *splitRegionsHandler) WatchRegionsByKeyRange(ctx context.Context, startKey, endKey []byte, splitKeys [][]byte,
+	timeout, watchInterval time.Duration, response *splitKeyResponse, wg *sync.WaitGroup) {
 	after := time.After(timeout)
 	ticker := time.NewTicker(watchInterval)
-	defer ticker.Stop()
-	regionsID := make(map[uint64]struct{}, len(splitKeys))
+	defer func() {
+		ticker.Stop()
+		wg.Done()
+	}()
 	for {
 		select {
 		case <-ticker.C:
 			regions := h.cluster.ScanRegions(startKey, endKey, -1)
 			for _, region := range regions {
 				for _, key := range splitKeys {
 					if bytes.Equal(key, region.GetStartKey()) {
-						regionsID[region.GetID()] = struct{}{}
+						log.Info("found split region",
+							zap.Uint64("region-id", region.GetID()),
+							logutil.ZapRedactString("split-key", hex.EncodeToString(key)))
+						response.addRegionsID(region)
 					}
 				}
 			}
-			if len(regionsID) < len(splitKeys) {
+			if len(response.getRegionsID()) < len(splitKeys) {
 				continue
 			}
-			return regionsID
+			return
 		case <-after:
-			return regionsID
+			return
+		case <-ctx.Done():
+			return
 		}
 	}
 }
+
+type batch struct {
+	regionID uint64
+	keys     [][]byte
+	region   *core.RegionInfo
+}
+
+func appendKeyBatches(batches []batch, region *core.RegionInfo, groupKeys [][]byte, limit int) []batch {
```

Comment:
Seem we don't need to add the limit into the parameters

Reply:
removed